### DISCUSSION TITLE
Additional qa deployment environment

### DISCUSSION THREAD

**Discussion by clivewalkden** [MAD-I-T/magento-actions#47] [General]
Hi,

Could you confirm if it's possible to add another stage with overriding settings and which files would need to be changed for example adding a qa stage deployment.

Thanks in advance.

**Comment by seyuf** [comment]
Hum, 

yep, you can add multiple staging and override the default scripts as [follows](https://github.com/MAD-I-T/magento-actions#to-override-the-files-in-default-scripts-and-config-directories-without-forking) or even better use the overriding technique to load custom scripts and/or config. You would only need to place the files in scripts or config directory as in the default overriding process and call you action matching file name e.g:

-    
       ├── .github
       │   └── workflows # directory where the workflows are found, see below for an example of main.yml 
       ├── README.md 
       ├── magento # directory where you Magento source files should go 
       └── scripts/my-new-action.sh 

- 
    ```
      name: 'step to execute my new action'
      if: always()
      uses: MAD-I-T/magento-actions@v3.19
      env:
        COMPOSER_AUTH: ${{secrets.COMPOSER_AUTH}}
      with:
        process: 'my-new-action'
        override_settings: 1
   ```
   
   As for the multiple environments management you can setup something like this :

-    ```
        name: 'deploy in QA testing env'
        uses: MAD-I-T/magento-actions@master
        env:
          COMPOSER_AUTH: ${{secrets.COMPOSER_AUTH}}
          BUCKET_COMMIT: bucket-commit-${{github.sha}}.tar.gz
          MYSQL_ROOT_PASSWORD: magento
          MYSQL_DATABASE: magento
          HOST_DEPLOY_PATH: ${{secrets.QA_TESTING_HOST_DEPLOY_PATH}}
          HOST_DEPLOY_PATH_BUCKET: ${{secrets.QA_TESTING_HOST_DEPLOY_PATH}}/bucket
          SSH_PRIVATE_KEY: ${{secrets.QA_TESTING_SSH_PRIVATE_KEY}}
          SSH_CONFIG: ${{secrets.QA_TESTING_CONFIG}}
          WRITE_USE_SUDO: false
        with:
          php: '7.1'
          process: 'deploy-staging'
          deployer: 'no-permission-check'
          
      - name: 'start zero downtime deploy preprod'
        if: steps.build.outcome == 'success'
        uses: MAD-I-T/magento-actions@master
        env:
          COMPOSER_AUTH: ${{secrets.COMPOSER_AUTH}}
          BUCKET_COMMIT: bucket-commit-${{github.sha}}.tar.gz
          MYSQL_ROOT_PASSWORD: magento
          MYSQL_DATABASE: magento
          HOST_DEPLOY_PATH: ${{secrets.STAGE_HOST_DEPLOY_PATH}}
          HOST_DEPLOY_PATH_BUCKET: ${{secrets.STAGE_HOST_DEPLOY_PATH}}/bucket
          SSH_PRIVATE_KEY: ${{secrets.STAGE_SSH_PRIVATE_KEY}}
          SSH_CONFIG: ${{secrets.STAGE_SSH_CONFIG}}
          WRITE_USE_SUDO: false
        with:
          php: '7.1'
          process: 'deploy-staging'
          deployer: 'no-permission-check'
          override_settings: 1     
          
      ```
      
If you need to change or add behavior of the default staging env, you can override files in config directory.
Or change the secrets variables values.
To conclude people are using this module in all kind of ways. Sometimes, i do spy on the used by links to steal some clever ways people are using this tool. Like [here](https://github.com/Alekseon/magento2-widget-forms/blob/master/.github/workflows/m2-actions-test.yml) e.g for third-party extension builders.
The action is very flexible and light we intend to keep it that way. If you have suggestions, though, don't hesitate. 
On the topic of QA testing, we're trying to implement [mftf](https://developer.adobe.com/commerce/testing/functional-testing-framework/). But i doubt that will be in public repo as it'll increase the overall complexity.

**Reply by clivewalkden** [reply]
Thanks, confirmed I have a working version of this working as expected now.
